### PR TITLE
Ds/integration

### PR DESCRIPTION
**Pull request by druzhkov-paul**

### CODE REVIEW

File: configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py [hunk 1]
```diff
@@ -0,0 +1,2 @@
+_base_ = './fcos_r50_caffe_fpn_gn-head_4x4_1x_coco.py'
```
Comment (druzhkov-paul) [inline configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py:1]: Where does this model come from?

File: configs/fcos/fcos_x101_64x4d_fpn_gn-head_mstrain_640-800_4x2_2x_coco.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+_base_ = './fcos_r50_caffe_fpn_gn-head_4x4_1x_coco.py'
```
Comment (druzhkov-paul) [inline configs/fcos/fcos_x101_64x4d_fpn_gn-head_mstrain_640-800_4x2_2x_coco.py:1]: Where does this model come from?

File: mmdet/apis/test.py [hunk 3]
```diff
@@ -96,10 +96,11 @@ def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False):
...
             # encode mask results
-            if isinstance(result[0], tuple):
-                result = [(bbox_results, encode_mask_results(mask_results))
-                          for bbox_results, mask_results in result]
-        results.extend(result)
+            if isinstance(result, tuple):
+                result = list(result)
+                result[1] = encode_mask_results(result[1])
+                result = tuple(result)
+        results.append(result)
```
Comment (druzhkov-paul) [inline mmdet/apis/test.py:103]: Why is this change needed? Should it be applied to `single_gpu_test` as well?

File: mmdet/core/anchor/anchor_generator.py [hunk 4]
```diff
@@ -252,19 +253,20 @@ def single_level_grid_anchors(self,
...
         shift_x = torch.arange(0, feat_w, device=device) * stride[0]
         shift_y = torch.arange(0, feat_h, device=device) * stride[1]
-
-        shift_xx, shift_yy = self._meshgrid(shift_x, shift_y)
-        shifts = torch.stack([shift_xx, shift_yy, shift_xx, shift_yy], dim=-1)
+        shift_yy, shift_xx = meshgrid(shift_y, shift_x)
+        shifts = torch.stack([shift_xx, shift_yy, shift_xx, shift_yy], dim=-1).view(-1, 1, 4)
+        # open-mmlab/mmdetection suggests the following shifts
+        #shift_xx, shift_yy = self._meshgrid(shift_x, shift_y)
+        #shifts = torch.stack([shift_xx, shift_yy, shift_xx, shift_yy], dim=-1)
```
Comment (druzhkov-paul) [inline mmdet/core/anchor/anchor_generator.py:264]: Clean up the commented code.

File: mmdet/core/anchor/anchor_generator.py [hunk 5]
```diff
@@ -322,10 +324,10 @@ def single_level_valid_flags(self,
         valid_y = torch.zeros(feat_h, dtype=torch.bool, device=device)
         valid_x[:valid_w] = 1
         valid_y[:valid_h] = 1
-        valid_xx, valid_yy = self._meshgrid(valid_x, valid_y)
+        valid_xx, valid_yy = meshgrid(valid_x, valid_y)
```
Comment (druzhkov-paul) [inline mmdet/core/anchor/anchor_generator.py:327]: Should not it be in reverse?

File: mmdet/core/bbox/assigners/atss_assigner.py [hunk 6]
```diff
@@ -119,9 +120,11 @@ def assign(self,
...
             end_idx = start_idx + bboxes_per_level
             distances_per_level = distances[start_idx:end_idx, :]
-            selectable_k = min(self.topk, bboxes_per_level)
-            _, topk_idxs_per_level = distances_per_level.topk(
-                selectable_k, dim=0, largest=False)
+            _, topk_idxs_per_level = topk(distances_per_level, self.topk, dim=0, largest=False)
+            # open-mmlab/mmdetection suggests the following topk using
+            #selectable_k = min(self.topk, bboxes_per_level)
+            #_, topk_idxs_per_level = distances_per_level.topk(
+            #    selectable_k, dim=0, largest=False)
```
Comment (druzhkov-paul) [inline mmdet/core/bbox/assigners/atss_assigner.py:127]: Just to clarify: the commented code (the one from upstream) does not work properly for export, right?

File: mmdet/core/bbox/coder/delta_xywh_bbox_coder.py [hunk 7]
```diff
@@ -4,6 +4,8 @@
 from ..builder import BBOX_CODERS
 from .base_bbox_coder import BaseBBoxCoder
 
+from mmdet.core.bbox.transforms import clamp
```
Comment (druzhkov-paul) [inline mmdet/core/bbox/coder/delta_xywh_bbox_coder.py:7]: Seems that either this import is redundant, or `Tensor.clamp` calls in this file should be replaced by this `clamp` calls.

File: mmdet/core/bbox/transforms.py [hunk 8]
```diff
@@ -1,5 +1,37 @@
 import numpy as np
 import torch
+from torch.onnx import is_in_onnx_export
+
+from ..utils.misc import to_numpy
+
+
+def clamp(x, min, max):
```
Comment (druzhkov-paul) [inline mmdet/core/bbox/transforms.py:8]: Do we still need this custom `clamp`, or we can rely on `Tensor.clamp` solely?

File: mmdet/core/post_processing/bbox_nms.py [hunk 9]
```diff
@@ -1,6 +1,16 @@
...
 import torch
-from mmcv.ops.nms import batched_nms
-
+import torch.onnx.symbolic_helper as sym_help
+from torch.autograd import Function
+from torch.onnx import is_in_onnx_export
+from torch.onnx.symbolic_opset9 import reshape
+from torch.onnx.symbolic_opset10 import _slice
+
+import mmdet.ops.nms  # Replace mmcv.ops.nms by one from mmdet.ops.nms
```
Comment (druzhkov-paul) [inline mmdet/core/post_processing/bbox_nms.py:9]: Why?

File: mmdet/core/post_processing/bbox_nms.py [hunk 10]
```diff
@@ -12,74 +22,86 @@ def multiclass_nms(multi_bboxes,
...
-        nms_thr (float): NMS IoU threshold
-        max_num (int, optional): if there are more than max_num bboxes after
-            NMS, only top max_num will be kept. Default to -1.
-        score_factors (Tensor, optional): The factors multiplied to scores
-            before applying NMS. Default to None.
+        nms_cfg (dict): NMS operation config
+        max_num (int): if there are more than max_num bboxes after NMS,
+            only top max_num will be kept.
+        score_factors (Tensor): The factors multiplied to scores before
+            applying NMS
```
Comment (druzhkov-paul) [inline mmdet/core/post_processing/bbox_nms.py:35]:
```suggestion
        max_num (int, optional): if there are more than max_num bboxes after
            NMS, only top max_num will be kept. Default to -1.
        score_factors (Tensor, optional): The factors multiplied to scores
            before applying NMS. Default to None.

```

File: mmdet/core/post_processing/bbox_nms.py [hunk 11]
```diff
@@ -1,6 +1,16 @@
+import sys
 import torch
-from mmcv.ops.nms import batched_nms
-
+import torch.onnx.symbolic_helper as sym_help
+from torch.autograd import Function
+from torch.onnx import is_in_onnx_export
+from torch.onnx.symbolic_opset9 import reshape
+from torch.onnx.symbolic_opset10 import _slice
```
Comment (druzhkov-paul) [inline mmdet/core/post_processing/bbox_nms.py:7]: remove?

File: mmdet/core/post_processing/bbox_nms.py [hunk 12]
```diff
@@ -1,6 +1,16 @@
+import sys
 import torch
-from mmcv.ops.nms import batched_nms
-
+import torch.onnx.symbolic_helper as sym_help
+from torch.autograd import Function
```
Comment (druzhkov-paul) [inline mmdet/core/post_processing/bbox_nms.py:4]: remove?

File: mmdet/datasets/pipelines/loading.py [hunk 13]
```diff
@@ -380,7 +391,9 @@ def __repr__(self):
         repr_str += f'with_mask={self.with_mask}, '
         repr_str += f'with_seg={self.with_seg}, '
         repr_str += f'poly2mask={self.poly2mask}, '
+        repr_str += f'with_text={self.with_text}, '
         repr_str += f'poly2mask={self.file_client_args})'
+        #repr_str += f'file_client_args={self.file_client_args})'
```
Comment (druzhkov-paul) [inline mmdet/datasets/pipelines/loading.py:396]: ?

File: mmdet/datasets/samplers/distributed_sampler.py [hunk 14]
```diff
@@ -20,6 +20,8 @@ def __iter__(self):
             indices = torch.arange(len(self.dataset)).tolist()
 
         # add extra samples to make it evenly divisible
+        while len(indices) < self.total_size:
+            indices += indices[:(self.total_size - len(indices))]
```
Comment (druzhkov-paul) [inline mmdet/datasets/samplers/distributed_sampler.py:24]: ?

File: mmdet/models/backbones/__init__.py [hunk 15]
```diff
@@ -1,3 +1,4 @@
+from .mobilenetv2 import SSDMobilenetV2
```
Comment (druzhkov-paul) [inline mmdet/models/backbones/__init__.py:1]: Do we need it?

File: mmdet/models/dense_heads/fcos_head.py [hunk 16]
```diff
@@ -412,6 +410,34 @@ def _get_points_single(self,
...
+    def _get_points_single(self, featmap_size, stride, dtype, device):
+        h, w = featmap_size
+        x_range = torch.arange(
+            0, w * stride, stride, dtype=dtype, device=device)
+        y_range = torch.arange(
+            0, h * stride, stride, dtype=dtype, device=device)
+        y, x = meshgrid(y_range, x_range)
+        points = torch.stack(
+            (x.reshape(-1), y.reshape(-1)), dim=-1) + stride // 2
+        return points
```
Comment (druzhkov-paul) [inline mmdet/models/dense_heads/fcos_head.py:440]: ?

File: mmdet/models/dense_heads/fovea_head.py [hunk 17]
```diff
@@ -120,6 +123,18 @@ def forward_single(self, x):
         cls_score = self.conv_cls(cls_feat)
         return cls_score, bbox_pred
 
+    def get_points(self, featmap_sizes, dtype, device, flatten=False):
```
Comment (druzhkov-paul) [inline mmdet/models/dense_heads/fovea_head.py:126]: ?

File: mmdet/models/detectors/base.py [hunk 18]
```diff
@@ -328,10 +401,24 @@ def show_result(self,
...
         if segm_result is not None and len(labels) > 0:  # non empty
             segms = mmcv.concat_list(segm_result)
-            if isinstance(segms[0], torch.Tensor):
-                segms = torch.stack(segms, dim=0).detach().cpu().numpy()
-            else:
-                segms = np.stack(segms, axis=0)
+            #if isinstance(segms[0], torch.Tensor):
+            #    segms = torch.stack(segms, dim=0).detach().cpu().numpy()
+            #else:
+            #    segms = np.stack(segms, axis=0)
```
Comment (druzhkov-paul) [inline mmdet/models/detectors/base.py:407]: ?

File: mmdet/models/roi_heads/htc_roi_head.py [hunk 19]
```diff
@@ -173,8 +176,11 @@ def _bbox_forward(self, stage, x, rois, semantic_feat=None):
             bbox_semantic_feat = self.semantic_roi_extractor([semantic_feat],
                                                              rois)
             if bbox_semantic_feat.shape[-2:] != bbox_feats.shape[-2:]:
-                bbox_semantic_feat = F.adaptive_avg_pool2d(
-                    bbox_semantic_feat, bbox_feats.shape[-2:])
+                if torch.onnx.is_in_onnx_export():
+                    bbox_semantic_feat = self.avg_pooling(bbox_semantic_feat)
+                else:
+                    bbox_semantic_feat = F.adaptive_avg_pool2d(
+                        bbox_semantic_feat, bbox_feats.shape[-2:])
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/htc_roi_head.py:183]: Are these alternatives equivalent?

File: mmdet/models/roi_heads/mask_heads/fcn_mask_head.py [hunk 20]
```diff
@@ -9,6 +9,9 @@
 
 from mmdet.core import mask_target
 from mmdet.models.builder import HEADS, build_loss
+from mmcv.ops import Conv2d
+from mmcv.ops.carafe import CARAFEPack
```
Comment (druzhkov-paul) [inline mmdet/models/roi_heads/mask_heads/fcn_mask_head.py:13]: Seems to be not used in modified code chunks.

File: tools/test.py [hunk 21]
```diff
@@ -169,51 +176,72 @@ def main():
...
+        kwargs['metric'] = args.eval
         if args.format_only:
             dataset.format_results(outputs, **kwargs)
         if args.eval:
-            eval_kwargs = cfg.get('evaluation', {}).copy()
-            # hard-code way to remove EvalHook args
-            for key in [
-                    'interval', 'tmpdir', 'start', 'gpu_collect', 'save_best',
-                    'rule'
-            ]:
```
Comment (druzhkov-paul) [inline tools/test.py:213]: I guess, all of these keys should be processed.

File: setup.py [hunk 22]
```diff
@@ -156,6 +155,25 @@ def gen_packages_items():
...
+                module='mmdet.ops.roi_align',
+                sources=[
+                    'src/roi_align_ext.cpp',
+                    'src/cpu/roi_align_v2.cpp',
+                ],
+                sources_cuda=[
+                    'src/cuda/roi_align_kernel.cu',
+                    'src/cuda/roi_align_kernel_v2.cu'
+                ]),
+        ],
```
Comment (druzhkov-paul) [inline setup.py:177]: What is the reason to still have them here?

File: requirements/optional.txt [hunk 23]
```diff
@@ -1,6 +1,10 @@
...
 imagecorruptions
+scikit-learn
+ninja
+tensorboard
+imagesize==1.2.0
+PyYAML==5.3.1
+sty==1.0.0b12
 mmlvis
-scipy
-sklearn
```
Comment (druzhkov-paul) [inline requirements/optional.txt:6]: ?

File: mmdet/utils/collect_env.py [hunk 24]
```diff
@@ -1,13 +1,84 @@
-from mmcv.utils import collect_env as collect_base_env
-from mmcv.utils import get_git_hash
+import os.path as osp
```
Comment (druzhkov-paul) [inline mmdet/utils/collect_env.py:1]: Update this file.

File: mmdet/utils/flops_counter.py [hunk 25]
```diff
@@ -0,0 +1,444 @@
+# Modified from flops-counter.pytorch by Vladislav Sovrasov
```
Comment (druzhkov-paul) [inline mmdet/utils/flops_counter.py:1]: I guess this has been moved to mmcv. Should we keep a copy?

File: configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py (hunk 1 above)
Comment (DmitriySidnev) [reply configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py:1]: Config files [fcos_r50_caffe_fpn_gn-head_4x4_1x_coco.py](https://github.com/openvinotoolkit/mmdetection/blob/ote/configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py), [fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py](https://github.com/openvinotoolkit/mmdetection/blob/ote/configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py) were added by us.

File: configs/fcos/fcos_x101_64x4d_fpn_gn-head_mstrain_640-800_4x2_2x_coco.py (hunk 2 above)
Comment (DmitriySidnev) [reply configs/fcos/fcos_x101_64x4d_fpn_gn-head_mstrain_640-800_4x2_2x_coco.py:1]: Config files [fcos_r50_caffe_fpn_gn-head_4x4_1x_coco.py](https://github.com/openvinotoolkit/mmdetection/blob/ote/configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py), [fcos_x101_64x4d_fpn_gn-head_mstrain_640-800_4x2_2x_coco.py](https://github.com/openvinotoolkit/mmdetection/blob/ote/configs/fcos/fcos_x101_64x4d_fpn_gn-head_mstrain_640-800_4x2_2x_coco.py) were added by us.

File: mmdet/apis/test.py (hunk 3 above)
Comment (DmitriySidnev) [reply mmdet/apis/test.py:103]: Yes, fixed as in `single_gpu_test `.

File: mmdet/core/anchor/anchor_generator.py (hunk 5 above)
Comment (DmitriySidnev) [reply mmdet/core/anchor/anchor_generator.py:327]: Not sure, [openvinotoolkit/mmdetection/ote/mmdet/core/anchor/anchor_generator.py](https://github.com/openvinotoolkit/mmdetection/blob/ote/mmdet/core/anchor/anchor_generator.py#L252).

File: mmdet/core/post_processing/bbox_nms.py (hunk 9 above)
Comment (DmitriySidnev) [reply mmdet/core/post_processing/bbox_nms.py:9]: With `nms` from `mmcv` export to ONNX does not work. Probably the reasons are not so complicated but I suggest to solve the problem with custom operations later independently from this PR.

File: mmdet/datasets/samplers/distributed_sampler.py (hunk 14 above)
Comment (DmitriySidnev) [reply mmdet/datasets/samplers/distributed_sampler.py:24]: [Commit](https://github.com/openvinotoolkit/mmdetection/commit/b3663907bb094ad8cac10830686fa4ab73654483#diff-63fba73f6ddba3db0fe9843e925e9335785a5a67225d2a525ed88f1c60a9d389). Ilya?

File: configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py (hunk 1 above)
Comment (druzhkov-paul) [reply configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py:1]: Don't think so: https://github.com/open-mmlab/mmdetection/blob/v2.8.0/configs/fcos/fcos_center_r50_caffe_fpn_gn-head_4x4_1x_coco.py . I believe that it was removed and readded back at some point, that's why the history don't show original authors. Please remove configs that are not used in upstream master anymore.

File: mmdet/datasets/samplers/distributed_sampler.py (hunk 14 above)
Comment (druzhkov-paul) [reply mmdet/datasets/samplers/distributed_sampler.py:24]: Ilya did it right. Alternative way of doing the same thing has been introduced in upstream (see a chunk of code right after this one), please leave only one way of doing this thing. I'd prefer the one from upstream to reduce the diff.

**Comment by druzhkov-paul** [comment]
I've moved the comments to openvinotoolkit/mmdetection. Let's continue conversation there.
